internal/hook/wait_exec_hook_handler.go

```diff
@@ -0,0 +1,263 @@
+}
+
+var _ ListWatchFactory = &DefaultListWatchFactory{}
+
+type DefaultWaitExecHookHandler struct {
+	ListWatchFactory   ListWatchFactory
+	PodCommandExecutor podexec.PodCommandExecutor
+}
+
+var _ WaitExecHookHandler = &DefaultWaitExecHookHandler{}
+
+func (e *DefaultWaitExecHookHandler) HandleHooks(
+	ctx context.Context,
+	log logrus.FieldLogger,
+	pod *v1.Pod,
+	byContainer map[string][]PodExecRestoreHook,
+) []error {
+	if pod == nil {
+		return nil
+	}
+
+	// If hooks are defined for a container that does not exist in the pod log a warning and discard
+	// those hooks to avoid waiting for a container that will never become ready. After that if
+	// there are no hooks left to be executed return immediately.
+	for containerName := range byContainer {
+		if !podHasContainer(pod, containerName) {
+			log.Warningf("Pod %s/%s does not have container %s: discarding post-restore exec hooks", pod.Namespace, pod.Name, containerName)
+			delete(byContainer, containerName)
+		}
+	}
+	if len(byContainer) == 0 {
+		return nil
+	}
+
+	// Every hook in every container can have its own wait timeout. Rather than setting up separate
+	// contexts for each, find the largest wait timeout for any hook that should be executed in
+	// the pod and watch the pod for up to that long. Before executing any hook in a container,
+	// check if that hook has a timeout and skip execution if expired.
+	ctx, cancel := context.WithCancel(ctx)
+	maxWait := maxHookWait(byContainer)
+	// If no hook has a wait timeout then this function will continue waiting for containers to
+	// become ready until the shared hook context is canceled.
+	if maxWait > 0 {
+		ctx, cancel = context.WithTimeout(ctx, maxWait)
+	}
+	waitStart := time.Now()
+
+	var errors []error
+
+	// The first time this handler is called after a container starts running it will execute all
+	// pending hooks for that container. Subsequent invocations of this handler will never execute
+	// hooks in that container. It uses the byContainer map to keep track of which containers have
+	// not yet been observed to be running. It relies on the Informer not to be called concurrently.
+	// When a container is observed running and its hooks are executed, the container is deleted
+	// from the byContainer map. When the map is empty the watch is ended.
+	handler := func(newObj interface{}) {
+		newPod, ok := newObj.(*v1.Pod)
+		if !ok {
+			return
+		}
+
+		if newPod.Status.Phase == v1.PodSucceeded || newPod.Status.Phase == v1.PodFailed {
+			err := fmt.Errorf("Pod %s/%s entered phase %s before some post-restore exec hooks ran", newPod.Namespace, newPod.Name, newPod.Status.Phase)
+			log.Warning(err)
+			cancel()
+			return
+		}
+
+		for containerName, hooks := range byContainer {
+			if !isContainerRunning(newPod, containerName) {
+				log.Infof("Container % in pod %s/%s is not running: post-restore hooks will not yet be executed", containerName, newPod.Namespace, newPod.Name)
```

Comment:
Can use [`namespaceAndName`](https://github.com/vmware-tanzu/velero/blob/f42c63af1b9af445e38f78a7256b1c48ef79c10e/pkg/util/kube/utils.go#L36) for the pod here.